clash-lib/src/Clash/Normalize/Transformations/ANF.hs

```diff
@@ -0,0 +1,334 @@
+--    depends on IO actions to be propagated down as far as possible.
+collectANF :: HasCallStack => NormRewriteW
+collectANF ctx e@(App appf arg)
+  | (conVarPrim, _) <- collectArgs e
+  , isCon conVarPrim || isPrim conVarPrim || isVar conVarPrim
+  = do
+    tcm <- Lens.view tcCache
+    untranslatable <- lift (isUntranslatable False arg)
+    let localVar = isLocalVar arg
+        constantNoCR = isConstantNotClockReset tcm arg
+    -- See Note [ANF no let-bind]
+    case (untranslatable,localVar || constantNoCR, isSimBind conVarPrim,arg) of
+      (False,False,False,_) -> do
+        -- See Note [ANF InScopeSet]
+        is1 <- Lens.use _2
+        argId <- lift (mkTmBinderFor is1 tcm (mkDerivedName ctx "app_arg") arg)
+        -- See Note [ANF InScopeSet]
+        tellBinders [(argId,arg)]
+        return (App appf (Var argId))
+      (True,False,_,Letrec binds body) -> do
+        tellBinders binds
+        return (App appf body)
+      _ -> return e
+ where
+  isSimBind (Prim p) = primName p == "Clash.Explicit.SimIO.bindSimIO#"
+  isSimBind _ = False
+
+collectANF _ (Letrec binds body) = do
+  tcm <- Lens.view tcCache
+  let isSimIO = isSimIOTy tcm (termType tcm body)
+  untranslatable <- lift (isUntranslatable False body)
+  let localVar = isLocalVar body
+  -- See Note [ANF no let-bind]
+  if localVar || untranslatable || isSimIO
+    then do
+      tellBinders binds
+      return body
+    else do
+      -- See Note [ANF InScopeSet]
+      is1 <- Lens.use _2
+      argId <- lift (mkTmBinderFor is1 tcm (mkUnsafeSystemName "result" 0) body)
+      -- See Note [ANF InScopeSet]
+      tellBinders [(argId,body)]
+      tellBinders binds
+      return (Var argId)
+
+-- TODO: The code below special-cases ANF for the ':-' constructor for the
+-- 'Signal' type. The 'Signal' type is essentially treated as a "transparent"
+-- type by the Clash compiler, so observing its constructor leads to all kinds
+-- of problems. In this case that "Clash.Rewrite.Util.mkSelectorCase" will
+-- try to project the LHS and RHS of the ':-' constructor, however,
+-- 'mkSelectorCase' uses 'coreView1' to find the "real" data-constructor.
+-- 'coreView1' however looks through the 'Signal' type, and hence 'mkSelector'
+-- finds the data constructors for the element type of Signal. This resulted in
+-- error #24 (https://github.com/christiaanb/clash2/issues/24), where we
+-- try to get the first field out of the 'Vec's 'Nil' constructor.
+--
+-- Ultimately we should stop treating Signal as a "transparent" type and deal
+-- handling of the Signal type, and the involved co-recursive functions,
+-- properly. At the moment, Clash cannot deal with this recursive type and the
+-- recursive functions involved, hence the need for special-casing code. After
+-- everything is done properly, we should remove the two lines below.
+collectANF _ e@(Case _ _ [(DataPat dc _ _,_)])
+  | nameOcc (dcName dc) == "Clash.Signal.Internal.:-" = return e
```

Comment:
```suggestion
  | nameOcc (dcName dc) == show '(:-) = return e
```

Reply:
> `show '(:-)`

Sorcery!

Reply:
I didn't know you could do that though, _neat_